src/main/java/com/google/sps/data/DeliverySlot.java

```diff
@@ -14,29 +14,94 @@
 package com.google.sps.data;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
-import java.util.Date; 
+import com.google.maps.errors.ApiException;
+import com.google.maps.model.LatLng;
+
+import java.io.IOException;
+import java.util.Date;
 import java.time.LocalTime;
 
-/** 
+/**
  * Class that represents a delivery timeslot using delivery date, start and end times.
  */
 public class DeliverySlot {
+  /** 
+   * Enum with the properties of a delivery slot in datastore.
+   */
+  public enum Property {
+    START_TIME("starTime"),
+    END_TIME("endTime"),
+    START_LAT("startLat"),
+    START_LNG("startLng"),
+    USER_ID("uid");
+
+    public final String label;
+
+    private Property(String label) {
+      this.label = label;
+    }
+  }
+
   private Date startTime;
   private Date endTime;
-  
-  /** 
-   * Creates a delivery slot given a date, and start and end times. 
+  private Point startPoint;
+  private String userId;
+  private String slotId;
+
+  /**
+   * Creates a delivery slot given a date, start and end times in miliseconds and the user id.
    */
-  public DeliverySlot(Date deliveryDay, long startTimeInMiliseconds, long endTimeInMiliseconds) {
+  public DeliverySlot(Date deliveryDay, long startTimeInMiliseconds, long endTimeInMiliseconds, String userId) throws BadRequestException {
+    if (startTimeInMiliseconds > endTimeInMiliseconds) {
+      throw new BadRequestException("Please enter a valid time!");
+    }
+    if (startTimeInMiliseconds < System.currentTimeMillis() - deliveryDay.getTime()) {
+      throw new BadRequestException("Please enter a valid date!");
+    }
     this.startTime = new Date(deliveryDay.getTime() + startTimeInMiliseconds);
     this.endTime = new Date(deliveryDay.getTime() + endTimeInMiliseconds);
+    this.userId = userId;
+  }
+
+  /**
+   * Given latitude and longitude, sets the coordinates for the starting point of the delivery.
+   */
+  public void setCoordinates(double latitude, double longitude) throws BadRequestException {
+    this.startPoint = new Point(latitude, longitude);
+  }
+
+  /**
+   * Given a string address, sets the coordinates for the starting point of the delivery.
+   */
+  public void setCoordinatesFromAddress(String startAddress) throws ApiException, BadRequestException, InterruptedException, IOException, DataNotFoundException {
```

Comment:
nit: I would just keep this a standard set method.

setStartPoint(Point point);

Then the caller can just call:

setStartPoint(Point.fromAddress(startAdress));

There is no reason this method cares about how the point is constructed really.

Reply:
I changed the methods to setStartPoint.